File: medium/55. Jump Game/55. Jump Game.py
from typing import List


class Solution:
    def canJump(self, nums: List[int]) -> bool:
        reachable = 0
        n = len(nums)

        for i, num in enumerate(nums):
            if i > reachable:
                return False

            reachable = max(reachable, i + num)

            if reachable >= n - 1:
                return True

        return False


solution = Solution()
print(solution.canJump([2, 3, 1, 1, 4]))
print(solution.canJump([3, 2, 1, 0, 4]))
print(solution.canJump([1, 0, 3, 1, 0, 0]))

File: medium/55. Jump Game/55. Jump Game.ts
function canJump(nums: number[]): boolean {
    const n = nums.length
    let max_reachable = 0

    for (let i = 0; i < n; i++) {
        const step = nums[i]
        if (i > step)
            return false

        max_reachable = Math.max(max_reachable, i + step)

        if (max_reachable >= n - 1)
            return true
    }

    return false
};